### ISSUE TITLE
Delay drawing till all tiles loaded

### ISSUE THREAD

**Issue by alxll** [openseadragon/openseadragon#736]
Hi,

There is a question - is it possible to delay drawing image untill all tile images loaded in memory?
I have to implement the following scenario:
1) user start the process of automatically changed tile source (we name it as "flight mode")
2) when the flight mode started the tile source changes by timer (e.g. each second or three seconds - configurable value). in my application it looks like that (angular):

scope.$watch('currentMedia', function(media) {
    scopeConfig.imageUrl(media.properties.url).then(function (url) {
        seadragonViewer.open(url);
     });
});

3) the number of images is unprecictable (potentially hundreds and thousands)
4) If tiles did not loaded during the period of the "exposing time" (one sec. for instance), the loading has to be canceled and new tile source loading has to be started while the seadragonViewer has to show the last successfully loaded image

At the moment tiles are showing in seadragonViewer independently after being loaded. From some point seadragonViewer cleared and shows nothing until the "flight mode" stops (and when it stops I have to wait for a few seconds untill all previously passed tile sources processed)

Is there any way to solve the descibed problem? any workarounds or hacks?

**Comment by iangilman** [comment]
Sounds like you have a seadragon-based slideshow, and you'd like to wait to switch images until the new image is fully loaded, is that correct? 

Unfortunately we do not yet have an event for "all tiles needed for this view loaded"; this would be good to add, and would be helpful in this situation. If you're interested in helping to add it, I can point you in the right direction. 

How are you switching between images? For a slide show, I would recommend loading the new image behind the old image and switching once the new image has loaded. We do  have an event for tile loads, so you can at least know when the first tile has loaded. You could then wait a brief moment to let a few more tiles load and then switch. 

Sounds like you're also having trouble with waiting for tiles to load even after you have moved on to the next image, correct? One way to handle this is to set the `imageLoaderLimit` option to something relatively small (4 maybe) when you create the viewer. It would probably also be good if the viewer cancelled all pending loads for any tiled image that was removed from the world...I don't believe we do that yet. Again, if you'd like to help out with that, I can point you in the right direction.

**Comment by alxll** [comment]
Hi,

Thanks a lot for the answer... but I propably provided not so good decription of my task :(

>>Sounds like you have a seadragon-based slideshow
mayby it may considered as a slideshow... but not sure that this is the best description...

>>and you'd like to wait to switch images until the new image is fully loaded, is that correct
No - this it totally incorrect - I would like to cancel loading previous image when starting a new image loading. The scenario you described I have implemented as (simplified and not optimized):

                var imageLoading = false;
                ....
                seadragonViewer.addHandler('open', function(){
                    imageLoading = false;
                });
                .............
                scope.$watch('currentMedia', function(media) {
                    scopeConfig.imageUrl(media.properties.url).then(function (url) {
                        if(!imageLoading) {
                            imageLoading = true;
                            seadragonViewer.open(url);
                        }
                    });
                });

But this is not the behaviour I need... I would like to have ability to do something like that (of course not exactly - thi is just the idea demonstration):
var tiledImageLoader;
scope.$watch('currentMedia', function(media) {
	if(tiledImageLoader && tiledImageLoader.status==='loading'){
		tiledImageLoader.abort();
	}
	tiledImageLoader = seadragonViewer.tiledImage(media.properties.url).then(function(tiledImage){
		tiledImage.show();
	});
});

In short - I neeed to cancel loading tiled image if it is not finished and start new loading...

>>How are you switching between images?
By timer... Some sequence number incremented in predefined time, then specific service call RESTFul service loading a media object which contains the actual URL. when the media object changed the code shown above executed. This behaviour is required to show aerial photos (original size ~30 Mb) emulating a flight on a helicopter... The helicopter may have many cameras (forward and backward cameras are typical example but it may have more then two cameras), each cameraman make a shot for instance each second or half a second (any interval is possible) 


>>I would recommend loading the new image behind the old image and switching once the new image has loaded
How can I do that? It looks promissing but I don't understand how I can implement it... Is it Ok to do that for 1000 images switching one by one?

And thanks again for your efforts to help me!

**Comment by iangilman** [comment]
Well, if you don't need to wait on one image loading before switching to the next image, it's probably not worth loading them both into the viewer at the same time. You certainly wouldn't want to do that with a 1000 images at the same time; I was just recommending 2 at the same time. 

At any rate, if you're just trying to abort in-flight loads, there's no mechanism in OSD for that, as it's not possible at all in JavaScript to abort an in-flight image load: 

http://stackoverflow.com/questions/14031281/javascript-to-cancel-image-loading

The next best thing is to use the `imageLoaderLimit` option so you limit how many loads can be in-flight at a time. This way when you switch images only those few will have to finish, and all others (that haven't started yet) will be discarded.

**Comment by alxll** [comment]
>>it's not possible at all in JavaScript to abort an in-flight image load

Please have a look on this:
http://plnkr.co/edit/YTfqAPiAUZKWbKqMj6CF?p=preview

The code is very ugly because it was written in 20 mins but you can see that it is possible (I was using this trick a few times in my job)

>>The next best thing is to use the imageLoaderLimit option

Ok, I will try that but for now I still confused because I don't understand the idea... Will investigate it...

**Comment by iangilman** [comment]
Interesting. The StackOverflow thread specifically said setting the img.src to an empty string didn't work. That said, it's a few years old, and your example certainly seems to be working, so I guess at least some of the browsers have fixed that issue. 

So, if you'd like to update OSD to support image cancellation, that would be in ImageLoader. We currently keep a queue of pending load jobs, but we'd now have to keep a new array of in-flight load jobs. When a new job is started: 

https://github.com/openseadragon/openseadragon/blob/master/src/imageloader.js#L136

...you'd stuff it in the in-flight array, and then on `clear()` you'd abort all of the in-flight jobs: 

https://github.com/openseadragon/openseadragon/blob/master/src/imageloader.js#L148

As for imageLoaderLimit, that works precisely because we can (and do) abort jobs that haven't started yet. If you specify a limit, anything over that limit goes into the queue. If you don't specify a limit, every job gets started as soon as it's created.

**Comment by alxll** [comment]
>> I guess at least some of the browsers have fixed that issue.

It works at least last 2 years with FF and chrome, mobile and desktop verstion and Safari at least for desktop... As ususal, there is no guaranties for IE. The StackOverflow  artical you are referencing describes the situation with <img> tag - there is a big difference (potentially - I don't know exactly) - <img> tag located in DOM tree when javascript change src...   

>>So, if you'd like to update OSD to support image cancellation, that would be in ImageLoader.

I gonna think about that but not now... First I will try to find some workaround - using imageLoaderLimit for instance or something else... Cancelable image loader required a Promisse implementation - you does not have jquery dependencies isn't it? so Promisses have to implemented from scratch - not a big deal but... for so small feature...

**Comment by iangilman** [comment]
I suppose we would need something like a Promise if you wanted to be able to cancel individual loads. There is already an API for cancelling all loads (`ImageLoader.clear()`), so that could be done with an array like I mentioned. 

At any rate, hopefully you can find a solution that works for you!

**Comment by alxll** [comment]
Yes - it is possible to do that as you describes but the result will be a bit ugly (but maybe I understood something wrong!) - imageLoader manages ImageJob array. ImageJob is a job which loads a single tile - isn't it? But the tile drawn immediately after loading (it looks like that)? so, if you abort all unfinished + none started jobs the image can be shown partially (one ore more tiles)? I would not say that this is a good behavior... 

Promisse is a good approach to solve it (show image when it is ready to show in current situation)... Not a single one, but very convinient from my point of view... 

And it is not absolutely required to use third party library for that - simplest promisses  can be implemented in plain javascript easily.. even more - it would better because classical promisses are non cancelable (you can so how stupidly I implemented the abort method in the example )

**Comment by iangilman** [comment]
I guess I'm not sure which tiles you're trying to abort then. I thought you wanted to abort old tile requests when switching to a new image. 

Anyway, sounds like there are options when we decide to pursue them.
